// @vendure/admin-ui/esm2022/settings/components/zone-member-list/zone-member-list-bulk-actions.mjs
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { DataService, NotificationService, Permission } from '@vendure/admin-ui/core';
export const removeZoneMembersBulkAction = {
    location: 'zone-members-list',
    label: _('settings.remove-from-zone'),
    icon: 'trash',
    iconClass: 'is-danger',
    requiresPermission: Permission.UpdateCustomerGroup,
    onClick: ({ injector, selection, hostComponent, clearSelection }) => {
        const dataService = injector.get(DataService);
        const notificationService = injector.get(NotificationService);
        const zone = hostComponent.activeZone;
        const memberIds = selection.map(s => s.id);
        dataService.settings.removeMembersFromZone(zone.id, memberIds).subscribe({
            complete: () => {
                notificationService.success(_(`settings.remove-countries-from-zone-success`), {
                    countryCount: memberIds.length,
                    zoneName: zone.name,
                });
                hostComponent.refresh();
                clearSelection();
            },
        });
    },
};
