// @vendure/admin-ui/esm2022/core/providers/guard/auth.guard.mjs
import { Injectable } from '@angular/core';
import { tap } from 'rxjs/operators';
import { getAppConfig } from '../../app.config';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "../auth/auth.service";
/**
 * This guard prevents unauthorized users from accessing any routes which require
 * authorization.
 */
export class AuthGuard {
    constructor(router, authService) {
        this.router = router;
        this.authService = authService;
        this.externalLoginUrl = getAppConfig().loginUrl;
    }
    canActivate(route) {
        return this.authService.checkAuthenticatedStatus().pipe(tap(authenticated => {
            if (!authenticated) {
                if (this.externalLoginUrl) {
                    window.location.href = this.externalLoginUrl;
                }
                else {
                    this.router.navigate(['/login']);
                }
            }
        }));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: AuthGuard, deps: [{ token: i1.Router }, { token: i2.AuthService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: AuthGuard, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: AuthGuard, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.Router }, { type: i2.AuthService }]; } });
