// @vendure/admin-ui/esm2022/core/shared/components/dropdown/dropdown-trigger.directive.mjs
import { Directive, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./dropdown.component";
export class DropdownTriggerDirective {
    constructor(dropdown, elementRef) {
        this.dropdown = dropdown;
        this.elementRef = elementRef;
        dropdown.setTriggerElement(this.elementRef);
    }
    onDropdownTriggerClick(event) {
        this.dropdown.toggleOpen();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DropdownTriggerDirective, deps: [{ token: i1.DropdownComponent }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.2", type: DropdownTriggerDirective, selector: "[vdrDropdownTrigger]", host: { listeners: { "click": "onDropdownTriggerClick($event)" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DropdownTriggerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[vdrDropdownTrigger]',
                }]
        }], ctorParameters: function () { return [{ type: i1.DropdownComponent }, { type: i0.ElementRef }]; }, propDecorators: { onDropdownTriggerClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
