// @vendure/admin-ui/esm2022/core/shared/components/dropdown/dropdown-item.directive.mjs
import { Directive, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./dropdown.component";
export class DropdownItemDirective {
    constructor(dropdown) {
        this.dropdown = dropdown;
    }
    onDropdownItemClick(event) {
        this.dropdown.onClick();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DropdownItemDirective, deps: [{ token: i1.DropdownComponent }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.2", type: DropdownItemDirective, selector: "[vdrDropdownItem]", host: { listeners: { "click": "onDropdownItemClick($event)" }, properties: { "class.dropdown-item": "true" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DropdownItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[vdrDropdownItem]',
                    // eslint-disable-next-line
                    host: { '[class.dropdown-item]': 'true' },
                }]
        }], ctorParameters: function () { return [{ type: i1.DropdownComponent }]; }, propDecorators: { onDropdownItemClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
