// @vendure/admin-ui/esm2022/core/shared/components/custom-detail-component-host/custom-detail-component-host.component.mjs
import { ChangeDetectionStrategy, Component, Injector, Input, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../../providers/custom-detail-component/custom-detail-component.service";
import * as i2 from "../ui-extension-point/ui-extension-point.component";
export class CustomDetailComponentHostComponent {
    constructor(viewContainerRef, customDetailComponentService, injector) {
        this.viewContainerRef = viewContainerRef;
        this.customDetailComponentService = customDetailComponentService;
        this.injector = injector;
        this.componentRefs = [];
    }
    ngOnInit() {
        const customComponents = this.customDetailComponentService.getCustomDetailComponentsFor(this.locationId);
        for (const config of customComponents) {
            const componentRef = this.viewContainerRef.createComponent(config.component, {
                injector: Injector.create({
                    parent: this.injector,
                    providers: config.providers ?? [],
                }),
            });
            componentRef.instance.entity$ = this.entity$;
            componentRef.instance.detailForm = this.detailForm;
            this.componentRefs.push(componentRef);
        }
    }
    ngOnDestroy() {
        for (const ref of this.componentRefs) {
            ref.destroy();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CustomDetailComponentHostComponent, deps: [{ token: i0.ViewContainerRef }, { token: i1.CustomDetailComponentService }, { token: i0.Injector }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.2", type: CustomDetailComponentHostComponent, selector: "vdr-custom-detail-component-host", inputs: { locationId: "locationId", entity$: "entity$", detailForm: "detailForm" }, ngImport: i0, template: "<vdr-ui-extension-point [locationId]=\"locationId\" api=\"detailComponent\"></vdr-ui-extension-point>\r\n", styles: [""], dependencies: [{ kind: "component", type: i2.UiExtensionPointComponent, selector: "vdr-ui-extension-point", inputs: ["locationId", "metadata", "topPx", "leftPx", "display", "api"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CustomDetailComponentHostComponent, decorators: [{
            type: Component,
            args: [{ selector: 'vdr-custom-detail-component-host', changeDetection: ChangeDetectionStrategy.OnPush, template: "<vdr-ui-extension-point [locationId]=\"locationId\" api=\"detailComponent\"></vdr-ui-extension-point>\r\n" }]
        }], ctorParameters: function () { return [{ type: i0.ViewContainerRef }, { type: i1.CustomDetailComponentService }, { type: i0.Injector }]; }, propDecorators: { locationId: [{
                type: Input
            }], entity$: [{
                type: Input
            }], detailForm: [{
                type: Input
            }] } });
