// @vendure/admin-ui/esm2022/core/providers/alerts/alerts.service.mjs
import { Injectable } from '@angular/core';
import { notNullOrUndefined } from '@vendure/common/lib/shared-utils';
import { BehaviorSubject, combineLatest, interval, isObservable, Subject, switchMap } from 'rxjs';
import { map, mapTo, startWith, take } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class Alert {
    constructor(config) {
        this.config = config;
        this.hasRun$ = new BehaviorSubject(false);
        this.data$ = new BehaviorSubject(undefined);
        if (this.config.recheckIntervalMs) {
            interval(this.config.recheckIntervalMs).subscribe(() => this.runCheck());
        }
        this.activeAlert$ = combineLatest(this.data$, this.hasRun$).pipe(map(([data, hasRun]) => {
            if (!data) {
                return;
            }
            const isAlert = this.config.isAlert(data);
            if (!isAlert) {
                return;
            }
            return {
                id: this.config.id,
                runAction: () => {
                    if (!hasRun) {
                        this.config.action(data);
                        this.hasRun$.next(true);
                    }
                },
                hasRun,
                label: this.config.label(data),
            };
        }));
    }
    get id() {
        return this.config.id;
    }
    runCheck() {
        const result = this.config.check();
        if (result instanceof Promise) {
            result.then(data => this.data$.next(data));
        }
        else if (isObservable(result)) {
            result.pipe(take(1)).subscribe(data => this.data$.next(data));
        }
        else {
            this.data$.next(result);
        }
        this.hasRun$.next(false);
    }
}
export class AlertsService {
    constructor() {
        this.alertsMap = new Map();
        this.configUpdated = new Subject();
        const alerts$ = this.configUpdated.pipe(mapTo([...this.alertsMap.values()]), startWith([...this.alertsMap.values()]));
        this.activeAlerts$ = alerts$.pipe(switchMap(() => {
            const alerts = [...this.alertsMap.values()];
            const isAlertStreams = alerts.map(alert => alert.activeAlert$);
            return combineLatest(isAlertStreams);
        }), map(alertStates => alertStates.filter(notNullOrUndefined)));
    }
    configureAlert(config) {
        this.alertsMap.set(config.id, new Alert(config));
        this.configUpdated.next();
    }
    refresh(id) {
        if (id) {
            this.alertsMap.get(id)?.runCheck();
        }
        else {
            this.alertsMap.forEach(config => config.runCheck());
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: AlertsService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: AlertsService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: AlertsService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return []; } });
