// @vendure/admin-ui/esm2022/core/extension/register-dashboard-widget.mjs
import { APP_INITIALIZER } from '@angular/core';
import { DashboardWidgetService } from '../providers/dashboard-widget/dashboard-widget.service';
/**
 * @description
 * Registers a dashboard widget. Once registered, the widget can be set as part of the default
 * (using {@link setDashboardWidgetLayout}).
 *
 * @docsCategory dashboard-widgets
 */
export function registerDashboardWidget(id, config) {
    return {
        provide: APP_INITIALIZER,
        multi: true,
        useFactory: (dashboardWidgetService) => () => {
            dashboardWidgetService.registerWidget(id, config);
        },
        deps: [DashboardWidgetService],
    };
}
/**
 * @description
 * Sets the default widget layout for the Admin UI dashboard.
 *
 * @docsCategory dashboard-widgets
 */
export function setDashboardWidgetLayout(layoutDef) {
    return {
        provide: APP_INITIALIZER,
        multi: true,
        useFactory: (dashboardWidgetService) => () => {
            dashboardWidgetService.setDefaultLayout(layoutDef);
        },
        deps: [DashboardWidgetService],
    };
}
