// @clr/angular/esm2020/progress/progress-bars/progress-bar.mjs
/*
 * Copyright (c) 2016-2023 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component, HostBinding, Input } from '@angular/core';
import { isBooleanAttributeSet } from '../../utils/component/is-boolean-attribute-set';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class ClrProgressBar {
    constructor() {
        this.max = 100;
        /*
         * No need to convert to `number` cause we could have
         * floating point and parseInt will round the numbers
         *
         * working with string won't have any side-effects,
         * we don't do any math so string will do the job.
         */
        this.value = 0;
        this.externalId = '';
    }
    get id() {
        return this._ID;
    }
    set id(value) {
        this._ID = value;
        this.externalId = null;
    }
    get progressClass() {
        return true;
    }
    set clrLabeled(value) {
        this._labeled = isBooleanAttributeSet(value);
    }
    get labeledClass() {
        return this._labeled;
    }
    set clrFade(value) {
        this._fade = isBooleanAttributeSet(value);
    }
    get fadeClass() {
        return this._fade;
    }
    set clrLoop(value) {
        this._loop = isBooleanAttributeSet(value);
    }
    get loopClass() {
        return this._loop;
    }
    /** @deprecated since 2.0, remove in 4.0 */
    set clrSuccess(value) {
        this._success = isBooleanAttributeSet(value);
    }
    get successClass() {
        return this._success;
    }
    /** @deprecated since 2.0, remove in 4.0 */
    set clrDanger(value) {
        this._danger = isBooleanAttributeSet(value);
    }
    get dangerClass() {
        return this._danger;
    }
    set clrFlash(value) {
        this._flash = isBooleanAttributeSet(value);
    }
    get flashClass() {
        return this._flash;
    }
    /** @deprecated since 2.0, remove in 4.0 */
    set clrFlashDanger(value) {
        this._flashDanger = isBooleanAttributeSet(value);
    }
    get flashDangerClass() {
        return this._flashDanger;
    }
    /**
     * Make sure that we always will have something that is readable
     * for the screen reader
     */
    get displayValue() {
        if (this.displayval) {
            return this.displayval;
        }
        return `${this.value || 0}%`;
    }
    /**
     * Display optional text only when labeled is eneabled
     */
    displayStringValue() {
        return this._labeled;
    }
}
ClrProgressBar.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.2", ngImport: i0, type: ClrProgressBar, deps: [], target: i0.ɵɵFactoryTarget.Component });
ClrProgressBar.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.2", type: ClrProgressBar, selector: "clr-progress-bar", inputs: { max: ["clrMax", "max"], displayval: ["clrDisplayval", "displayval"], value: ["clrValue", "value"], id: "id", clrLabeled: "clrLabeled", clrFade: "clrFade", clrLoop: "clrLoop", clrSuccess: "clrSuccess", clrDanger: "clrDanger", clrFlash: "clrFlash", clrFlashDanger: "clrFlashDanger" }, host: { properties: { "attr.id": "this.externalId", "class.progress": "this.progressClass", "class.labeled": "this.labeledClass", "class.progress-fade": "this.fadeClass", "class.loop": "this.loopClass", "class.success": "this.successClass", "class.danger": "this.dangerClass", "class.flash": "this.flashClass", "class.flash-danger": "this.flashDangerClass" } }, ngImport: i0, template: `
    <progress [id]="id" [attr.max]="max" [attr.value]="value" [attr.data-displayval]="displayValue"></progress>
    <span *ngIf="displayStringValue()">{{ displayValue }}</span>
  `, isInline: true, dependencies: [{ kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.2", ngImport: i0, type: ClrProgressBar, decorators: [{
            type: Component,
            args: [{
                    selector: 'clr-progress-bar',
                    template: `
    <progress [id]="id" [attr.max]="max" [attr.value]="value" [attr.data-displayval]="displayValue"></progress>
    <span *ngIf="displayStringValue()">{{ displayValue }}</span>
  `,
                }]
        }], propDecorators: { max: [{
                type: Input,
                args: ['clrMax']
            }], displayval: [{
                type: Input,
                args: ['clrDisplayval']
            }], value: [{
                type: Input,
                args: ['clrValue']
            }], externalId: [{
                type: HostBinding,
                args: ['attr.id']
            }], id: [{
                type: Input
            }], progressClass: [{
                type: HostBinding,
                args: ['class.progress']
            }], clrLabeled: [{
                type: Input,
                args: ['clrLabeled']
            }], labeledClass: [{
                type: HostBinding,
                args: ['class.labeled']
            }], clrFade: [{
                type: Input,
                args: ['clrFade']
            }], fadeClass: [{
                type: HostBinding,
                args: ['class.progress-fade']
            }], clrLoop: [{
                type: Input,
                args: ['clrLoop']
            }], loopClass: [{
                type: HostBinding,
                args: ['class.loop']
            }], clrSuccess: [{
                type: Input,
                args: ['clrSuccess']
            }], successClass: [{
                type: HostBinding,
                args: ['class.success']
            }], clrDanger: [{
                type: Input,
                args: ['clrDanger']
            }], dangerClass: [{
                type: HostBinding,
                args: ['class.danger']
            }], clrFlash: [{
                type: Input,
                args: ['clrFlash']
            }], flashClass: [{
                type: HostBinding,
                args: ['class.flash']
            }], clrFlashDanger: [{
                type: Input,
                args: ['clrFlashDanger']
            }], flashDangerClass: [{
                type: HostBinding,
                args: ['class.flash-danger']
            }] } });
