// @vendure/admin-ui/esm2022/login/login.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { SharedModule } from '@vendure/admin-ui/core';
import { LoginComponent } from './components/login/login.component';
import { loginRoutes } from './login.routes';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export class LoginModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: LoginModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.2", ngImport: i0, type: LoginModule, declarations: [LoginComponent], imports: [SharedModule, i1.RouterModule] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: LoginModule, imports: [SharedModule, RouterModule.forChild(loginRoutes)] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: LoginModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [SharedModule, RouterModule.forChild(loginRoutes)],
                    exports: [],
                    declarations: [LoginComponent],
                }]
        }] });
