// @vendure/admin-ui/esm2022/core/providers/custom-history-entry-component/history-entry-component.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class HistoryEntryComponentService {
    constructor() {
        this.customEntryComponents = new Map();
    }
    registerComponent(config) {
        this.customEntryComponents.set(config.type, config);
    }
    getComponent(type) {
        return this.customEntryComponents.get(type)?.component;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: HistoryEntryComponentService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: HistoryEntryComponentService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: HistoryEntryComponentService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
