// @vendure/admin-ui/esm2022/core/shared/components/split-view/split-view.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class SplitViewLeftDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: SplitViewLeftDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.2", type: SplitViewLeftDirective, selector: "[vdrSplitViewLeft]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: SplitViewLeftDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[vdrSplitViewLeft]',
                }]
        }] });
export class SplitViewRightDirective {
    constructor(template) {
        this.template = template;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: SplitViewRightDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.2", type: SplitViewRightDirective, selector: "[vdrSplitViewRight]", inputs: { splitViewTitle: "splitViewTitle" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: SplitViewRightDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[vdrSplitViewRight]',
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; }, propDecorators: { splitViewTitle: [{
                type: Input
            }] } });
