// @vendure/admin-ui/esm2022/core/shared/components/data-table-2/data-table-custom-component.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class DataTableCustomComponentService {
    constructor() {
        this.configMap = new Map();
    }
    registerCustomComponent(config) {
        const id = this.compoundId(config.tableId, config.columnId);
        this.configMap.set(id, config);
    }
    getCustomComponentsFor(tableId, columnId) {
        return this.configMap.get(this.compoundId(tableId, columnId));
    }
    compoundId(tableId, columnId) {
        return `${tableId}.${columnId}`;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DataTableCustomComponentService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DataTableCustomComponentService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DataTableCustomComponentService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
