// @vendure/admin-ui/esm2022/core/shared/components/data-table-2/data-table-column.component.mjs
import { Component, ContentChild, Input, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
export class DataTable2ColumnComponent {
    constructor() {
        /**
         * When set to true, this column will expand to use available width
         */
        this.expand = false;
        this.align = 'left';
        this.optional = true;
        this.hiddenByDefault = false;
        this.orderable = true;
        this.#visible = true;
        this.#onColumnChangeFns = [];
    }
    #visible;
    #onColumnChangeFns;
    get visible() {
        return this.#visible;
    }
    ngOnInit() {
        this.#visible = this.hiddenByDefault ? false : true;
    }
    setVisibility(isVisible) {
        this.#visible = isVisible;
        this.#onColumnChangeFns.forEach(fn => fn());
    }
    resetVisibility() {
        this.setVisibility(!this.hiddenByDefault);
    }
    onColumnChange(callback) {
        this.#onColumnChangeFns.push(callback);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DataTable2ColumnComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.2", type: DataTable2ColumnComponent, selector: "vdr-dt2-column", inputs: { id: "id", expand: "expand", heading: "heading", align: "align", sort: "sort", optional: "optional", hiddenByDefault: "hiddenByDefault", orderable: "orderable" }, queries: [{ propertyName: "template", first: true, predicate: TemplateRef, descendants: true }], exportAs: ["row"], ngImport: i0, template: ``, isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DataTable2ColumnComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'vdr-dt2-column',
                    template: ``,
                    exportAs: 'row',
                }]
        }], propDecorators: { id: [{
                type: Input
            }], expand: [{
                type: Input
            }], heading: [{
                type: Input
            }], align: [{
                type: Input
            }], sort: [{
                type: Input
            }], optional: [{
                type: Input
            }], hiddenByDefault: [{
                type: Input
            }], orderable: [{
                type: Input
            }], template: [{
                type: ContentChild,
                args: [TemplateRef, { static: false }]
            }] } });
