// @vendure/admin-ui/esm2022/order/providers/routing/order-resolver.mjs
import { Injectable } from '@angular/core';
import { ActivationStart } from '@angular/router';
import { notNullOrUndefined } from '@vendure/common/lib/shared-utils';
import { EMPTY } from 'rxjs';
import { filter, map, shareReplay, switchMap, take, takeUntil } from 'rxjs/operators';
import { DraftOrderDetailComponent } from '../../components/draft-order-detail/draft-order-detail.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "@vendure/admin-ui/core";
/**
 * Resolves the id from the path into a Customer entity.
 */
export class OrderResolver {
    constructor(router, dataService) {
        this.router = router;
        this.dataService = dataService;
    }
    /** @internal */
    resolve(route, state) {
        const id = route.paramMap.get('id');
        // Complete the entity stream upon navigating away
        const navigateAway$ = this.router.events.pipe(filter(event => event instanceof ActivationStart));
        const stream = this.dataService.order
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            .getOrder(id)
            .mapStream(data => data.order)
            .pipe(switchMap(order => {
            if (order?.state === 'Draft' && route.component !== DraftOrderDetailComponent) {
                // Make sure Draft orders only get displayed with the DraftOrderDetailComponent
                this.router.navigate(['/orders/draft', id]);
                return EMPTY;
            }
            else {
                return [order];
            }
        }), takeUntil(navigateAway$), filter(notNullOrUndefined), shareReplay(1));
        return stream.pipe(take(1), map(() => stream));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: OrderResolver, deps: [{ token: i1.Router }, { token: i2.DataService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: OrderResolver, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: OrderResolver, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.Router }, { type: i2.DataService }]; } });
