// @clr/angular/esm2020/button/providers/button-group-focus-handler.service.mjs
/*
 * Copyright (c) 2016-2023 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Injectable } from '@angular/core';
import { Linkers } from '../../utils/focus/focusable-item/linkers';
import { InitialFocus } from './button-group-focus.enum';
import * as i0 from "@angular/core";
import * as i1 from "../../utils/focus/focus.service";
import * as i2 from "../../utils";
export class ButtonGroupFocusHandler {
    constructor(focusService, toggleService, renderer) {
        this.focusService = focusService;
        this.toggleService = toggleService;
        this.renderer = renderer;
        this.initialFocus = InitialFocus.FIRST_ITEM;
        this._unlistenFuncs = [];
    }
    ngOnDestroy() {
        this._unlistenFuncs.forEach((unlisten) => unlisten());
        this.focusService.detachListeners();
    }
    initialize({ menu, menuToggle }) {
        this.menu = menu;
        this.menuToggle = menuToggle;
        this.focusService.registerContainer(this.menu, '-1');
        this.listenToKeys();
        this.linkButtons();
        switch (this.initialFocus) {
            case InitialFocus.LAST_ITEM:
                this.focusLastItem();
                break;
            default:
                this.focusFirstItem();
                break;
        }
    }
    resetButtonsFocus() {
        this.buttons.forEach(button => {
            button.blur();
        });
    }
    listenToKeys() {
        this._unlistenFuncs.push(this.renderer.listen(this.menu, 'keydown.shift.tab', event => this.closeMenu(event, false)));
        this._unlistenFuncs.push(this.renderer.listen(this.menu, 'keydown.tab', event => this.closeMenu(event, true)));
    }
    closeMenu(event, focusBackOnToggle) {
        this.toggleService.toggleWithEvent(event);
        if (focusBackOnToggle) {
            this.menuToggle.focus();
        }
        this.resetButtonsFocus();
    }
    linkButtons() {
        const buttonElements = Array.from(this.menu.children);
        this.buttons = buttonElements.map(buttonElement => {
            this._unlistenFuncs.push(this.renderer.listen(buttonElement, 'click', event => this.closeMenu(event, true)));
            return {
                id: buttonElement.id,
                value: buttonElement,
                focus: () => {
                    buttonElement.setAttribute('tabindex', '0');
                    buttonElement.focus();
                },
                blur: () => {
                    buttonElement.setAttribute('tabindex', '-1');
                    buttonElement.blur();
                },
            };
        });
        this.resetButtonsFocus();
        Linkers.linkVertical(this.buttons);
    }
    focusFirstItem() {
        if (this.buttons.length) {
            this.focusService.moveTo(this.buttons[0]);
        }
        this.initialFocus = InitialFocus.FIRST_ITEM;
    }
    focusLastItem() {
        if (this.buttons.length) {
            this.focusService.moveTo(this.buttons[this.buttons.length - 1]);
        }
        this.initialFocus = InitialFocus.FIRST_ITEM;
    }
}
ButtonGroupFocusHandler.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.2", ngImport: i0, type: ButtonGroupFocusHandler, deps: [{ token: i1.FocusService }, { token: i2.ClrPopoverToggleService }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Injectable });
ButtonGroupFocusHandler.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.2", ngImport: i0, type: ButtonGroupFocusHandler });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.2", ngImport: i0, type: ButtonGroupFocusHandler, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.FocusService }, { type: i2.ClrPopoverToggleService }, { type: i0.Renderer2 }]; } });
export const BUTTON_GROUP_FOCUS_HANDLER_PROVIDER = {
    provide: ButtonGroupFocusHandler,
};
