// @vendure/admin-ui/esm2022/core/shared/pipes/custom-field-label.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Displays a localized label for a CustomField or StringFieldOption, falling back to the
 * name/value if none are defined.
 */
export class CustomFieldLabelPipe {
    transform(value, uiLanguageCode) {
        if (!value) {
            return value;
        }
        if (Array.isArray(value)) {
            const match = value.find(l => l.languageCode === uiLanguageCode);
            return match ? match.value : value[0].value;
        }
        const { label } = value;
        const name = this.isCustomFieldConfig(value) ? value.name : value.value;
        if (label) {
            const match = label.find(l => l.languageCode === uiLanguageCode);
            return match ? match.value : label[0].value;
        }
        else {
            return name;
        }
    }
    isCustomFieldConfig(input) {
        return input.hasOwnProperty('name');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CustomFieldLabelPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.2", ngImport: i0, type: CustomFieldLabelPipe, name: "customFieldLabel" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CustomFieldLabelPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'customFieldLabel',
                    pure: true,
                }]
        }] });
