// @vendure/admin-ui/esm2022/core/shared/pipes/custom-field-description.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Displays a localized description for a CustomField
 */
export class CustomFieldDescriptionPipe {
    transform(value, uiLanguageCode) {
        if (!value) {
            return value;
        }
        const { description } = value;
        if (description) {
            const match = description.find(l => l.languageCode === uiLanguageCode);
            return match ? match.value : description[0].value;
        }
        else {
            return '';
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CustomFieldDescriptionPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "16.2.2", ngImport: i0, type: CustomFieldDescriptionPipe, name: "customFieldDescription" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CustomFieldDescriptionPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'customFieldDescription',
                    pure: true,
                }]
        }] });
