// @vendure/admin-ui/esm2022/core/shared/providers/routing/can-deactivate-detail-guard.mjs
import { Injectable } from '@angular/core';
import { marker as _ } from '@biesbjerg/ngx-translate-extract-marker';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "../../../providers/modal/modal.service";
import * as i2 from "@angular/router";
export class CanDeactivateDetailGuard {
    constructor(modalService, router) {
        this.modalService = modalService;
        this.router = router;
    }
    canDeactivate(component, currentRoute, currentState, nextState) {
        if (!component.canDeactivate()) {
            return this.modalService
                .dialog({
                title: _('common.confirm-navigation'),
                body: _('common.there-are-unsaved-changes'),
                buttons: [
                    { type: 'danger', label: _('common.discard-changes'), returnValue: true },
                    { type: 'primary', label: _('common.cancel-navigation'), returnValue: false },
                ],
            })
                .pipe(map(result => !!result));
        }
        else {
            return true;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CanDeactivateDetailGuard, deps: [{ token: i1.ModalService }, { token: i2.Router }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CanDeactivateDetailGuard }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CanDeactivateDetailGuard, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.ModalService }, { type: i2.Router }]; } });
