// @vendure/admin-ui/esm2022/core/shared/components/localized-text/localized-text.component.mjs
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../../data/providers/data.service";
import * as i2 from "@angular/common";
import * as i3 from "@ngx-translate/core";
import * as i4 from "../../pipes/custom-field-label.pipe";
export class LocalizedTextComponent {
    constructor(dataService) {
        this.dataService = dataService;
        this.uiLanguage$ = this.dataService.client.uiState().mapStream(data => data.uiState.language);
    }
    isString(value) {
        return typeof value === 'string';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: LocalizedTextComponent, deps: [{ token: i1.DataService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.2", type: LocalizedTextComponent, selector: "vdr-localized-text", inputs: { text: "text" }, ngImport: i0, template: "<ng-container *ngIf=\"isString(text)\">{{ text | translate }}</ng-container>\r\n<ng-container *ngIf=\"!isString(text)\">{{ text | customFieldLabel : (uiLanguage$ | async) }}</ng-container>\r\n", styles: [""], dependencies: [{ kind: "directive", type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "pipe", type: i2.AsyncPipe, name: "async" }, { kind: "pipe", type: i3.TranslatePipe, name: "translate" }, { kind: "pipe", type: i4.CustomFieldLabelPipe, name: "customFieldLabel" }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: LocalizedTextComponent, decorators: [{
            type: Component,
            args: [{ selector: 'vdr-localized-text', changeDetection: ChangeDetectionStrategy.OnPush, template: "<ng-container *ngIf=\"isString(text)\">{{ text | translate }}</ng-container>\r\n<ng-container *ngIf=\"!isString(text)\">{{ text | customFieldLabel : (uiLanguage$ | async) }}</ng-container>\r\n" }]
        }], ctorParameters: function () { return [{ type: i1.DataService }]; }, propDecorators: { text: [{
                type: Input
            }] } });
