// @vendure/admin-ui/esm2022/core/shared/directives/disabled.directive.mjs
import { Directive, Input, Optional } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/forms";
/**
 * Allows declarative binding to the "disabled" property of a reactive form
 * control.
 */
export class DisabledDirective {
    set disabled(val) {
        const formControl = this.formControlName?.control ?? this.formControl?.form;
        if (!formControl) {
            return;
        }
        if (!!val === false) {
            formControl.enable({ emitEvent: false });
        }
        else {
            formControl.disable({ emitEvent: false });
        }
    }
    constructor(formControlName, formControl) {
        this.formControlName = formControlName;
        this.formControl = formControl;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DisabledDirective, deps: [{ token: i1.FormControlName, optional: true }, { token: i1.FormControlDirective, optional: true }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.2", type: DisabledDirective, selector: "[vdrDisabled]", inputs: { disabled: ["vdrDisabled", "disabled"] }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DisabledDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[vdrDisabled]',
                }]
        }], ctorParameters: function () { return [{ type: i1.FormControlName, decorators: [{
                    type: Optional
                }] }, { type: i1.FormControlDirective, decorators: [{
                    type: Optional
                }] }]; }, propDecorators: { disabled: [{
                type: Input,
                args: ['vdrDisabled']
            }] } });
