// @vendure/admin-ui/esm2022/core/extension/components/angular-route.component.mjs
import { Component, inject } from '@angular/core';
import { SharedModule } from '../../shared/shared.module';
import { ROUTE_COMPONENT_OPTIONS, RouteComponent } from './route.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class AngularRouteComponent {
    constructor() {
        this.component = inject(ROUTE_COMPONENT_OPTIONS).component;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: AngularRouteComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.2", type: AngularRouteComponent, isStandalone: true, selector: "vdr-angular-route-component", ngImport: i0, template: ` <vdr-route-component><ng-container *ngComponentOutlet="component" /></vdr-route-component> `, isInline: true, dependencies: [{ kind: "ngmodule", type: SharedModule }, { kind: "directive", type: i1.NgComponentOutlet, selector: "[ngComponentOutlet]", inputs: ["ngComponentOutlet", "ngComponentOutletInputs", "ngComponentOutletInjector", "ngComponentOutletContent", "ngComponentOutletNgModule", "ngComponentOutletNgModuleFactory"] }, { kind: "component", type: RouteComponent, selector: "vdr-route-component" }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: AngularRouteComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'vdr-angular-route-component',
                    template: ` <vdr-route-component><ng-container *ngComponentOutlet="component" /></vdr-route-component> `,
                    standalone: true,
                    imports: [SharedModule, RouteComponent],
                }]
        }] });
