// @vendure/admin-ui/esm2022/catalog/components/collection-tree/collection-tree.service.mjs
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * Facilitates communication between the CollectionTreeComponent and child CollectionTreeNodeComponents
 * without introducing a cyclic dependency.
 */
export class CollectionTreeService {
    constructor() {
        this.allMoveListItems = [];
        this._rearrange$ = new Subject();
        this._delete$ = new Subject();
        this.rearrange$ = this._rearrange$.asObservable();
        this.delete$ = this._delete$.asObservable();
    }
    ngOnDestroy() {
        this._rearrange$.complete();
        this._delete$.complete();
    }
    setCollectionTree(tree) {
        this.collectionTree = tree;
    }
    resetMoveList() {
        this.allMoveListItems = [];
    }
    getMoveListItems(collection) {
        if (this.allMoveListItems.length === 0) {
            this.allMoveListItems = this.calculateAllMoveListItems();
        }
        return this.allMoveListItems.filter(item => item.id !== collection.id &&
            !item.ancestorIdPath.has(collection.id) &&
            item.id !== collection.parent?.id);
    }
    onDrop(event) {
        const item = event.item.data;
        const newParent = event.container.data;
        const newParentId = newParent.id;
        if (newParentId == null) {
            throw new Error(`Could not determine the ID of the root Collection`);
        }
        this._rearrange$.next({
            collectionId: item.id,
            parentId: newParentId,
            index: event.currentIndex,
        });
    }
    onMove(event) {
        this._rearrange$.next(event);
    }
    onDelete(id) {
        this._delete$.next(id);
    }
    calculateAllMoveListItems() {
        const visit = (node, parentPath, ancestorIdPath, output) => {
            const path = parentPath.concat(node.name);
            output.push({ path: path.slice(1).join(' / ') || 'root', id: node.id, ancestorIdPath });
            node.children.forEach(child => visit(child, path, new Set([...ancestorIdPath, node.id]), output));
            return output;
        };
        return visit(this.collectionTree, [], new Set(), []);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CollectionTreeService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CollectionTreeService }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CollectionTreeService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return []; } });
