// @vendure/admin-ui/esm2022/core/providers/custom-detail-component/custom-detail-component.service.mjs
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class CustomDetailComponentService {
    constructor() {
        this.customDetailComponents = new Map();
    }
    registerCustomDetailComponent(config) {
        if (this.customDetailComponents.has(config.locationId)) {
            this.customDetailComponents.get(config.locationId)?.push(config);
        }
        else {
            this.customDetailComponents.set(config.locationId, [config]);
        }
    }
    getCustomDetailComponentsFor(locationId) {
        return this.customDetailComponents.get(locationId) ?? [];
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CustomDetailComponentService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CustomDetailComponentService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: CustomDetailComponentService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
