// @vendure/admin-ui/esm2022/core/shared/components/rich-text-editor/prosemirror/context-menu/context-menu.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject, combineLatest, interval, of, Subject } from 'rxjs';
import { bufferWhen, debounceTime, delayWhen, distinctUntilChanged, filter, map, takeUntil, } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class ContextMenuService {
    constructor() {
        this.menuIsVisible$ = new BehaviorSubject(false);
        this.setContextMenuConfig$ = new Subject();
        const source$ = this.setContextMenuConfig$.asObservable();
        const groupedConfig$ = source$.pipe(bufferWhen(() => source$.pipe(debounceTime(50))), map(group => group.reduce((acc, cur) => {
            if (!acc) {
                return cur;
            }
            else {
                if (cur?.ref === acc.ref) {
                    acc.items.push(
                    // de-duplicate items
                    ...(cur?.items.filter(i => !acc.items.find(ai => ai.label === i.label)) ??
                        []));
                }
            }
            return acc;
        }, undefined)));
        const visible$ = this.menuIsVisible$.pipe(filter(val => val === true));
        const isVisible$ = this.menuIsVisible$.pipe(delayWhen(value => (value ? of(value) : interval(250).pipe(takeUntil(visible$)))), distinctUntilChanged());
        this.contextMenu$ = combineLatest(groupedConfig$, isVisible$).pipe(map(([config, isVisible]) => (isVisible ? config : undefined)));
    }
    setVisibility(isVisible) {
        this.menuIsVisible$.next(isVisible);
    }
    setContextMenu(config) {
        this.setContextMenuConfig$.next(config);
    }
    clearContextMenu() {
        this.setContextMenuConfig$.next(undefined);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: ContextMenuService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: ContextMenuService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: ContextMenuService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return []; } });
