// ngx-notify/esm2015/notify.js
import { ApplicationRef, Component, ComponentFactoryResolver, EventEmitter, Injectable, Injector, Input, NgModule, NgZone, Optional, ViewEncapsulation } from '@angular/core';
import { Observable, Subject } from 'rxjs';
import { DomSanitizer } from '@angular/platform-browser';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */

const defaultIcons = {
    alert: `
    <svg xmlns="http://www.w3.org/2000/svg" fill="#000000" height="24" viewBox="0 0 24 24" width="24">
      <path d="M0 0h24v24H0z" fill="none"/>
      <path d="M22 5.72l-4.6-3.86-1.29 1.53 4.6 3.86L22 5.72zM7.88 3.39L6.6 1.86 2 5.71l1.29 1.53 4.59-3.85zM12.5 8H11v6l4.75 2.85.75-1.23-4-2.37V8zM12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9c4.97 0 9-4.03 9-9s-4.03-9-9-9zm0 16c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"/>
    </svg>`,
    error: `
    <svg viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" width="24" height="24"><path d="M512 937.19552c-234.83392 0-425.19552-190.36672-425.19552-425.19552 0-234.8288 190.36672-425.19552 425.19552-425.19552 234.8288 0 425.20064 190.36672 425.20064 425.19552S746.8288 937.19552 512 937.19552z m0-797.24544c-205.47072 0-372.04992 166.57408-372.04992 372.04992 0 205.47072 166.5792 372.0448 372.04992 372.0448s372.0448-166.57408 372.0448-372.0448S717.47072 139.95008 512 139.95008z m37.7344 372.18816l112.59392 112.61952c10.37824 10.36288 10.37824 27.19744 0 37.57568s-27.20768 10.37824-37.57568 0l-112.61952-112.59392-113.3568 113.3568a26.75712 26.75712 0 0 1-37.87264 0 26.76224 26.76224 0 0 1 0-37.86752L474.2656 511.8464 361.66656 399.24736a26.5728 26.5728 0 1 1 37.5808-37.5808L511.8464 474.2656l114.20672-114.20672a26.78784 26.78784 0 0 1 37.87264 37.888l-114.19136 114.19136z"></path></svg>
    `,
    info: `
    <svg xmlns="http://www.w3.org/2000/svg" fill="#000000" height="24" viewBox="0 0 24 24" width="24">
      <path d="M0 0h24v24H0z" fill="none"/>
      <path d="M11 17h2v-6h-2v6zm1-15C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zM11 9h2V7h-2v2z"/>
    </svg>
    `,
    success: `
    <svg xmlns="http://www.w3.org/2000/svg" fill="#000000" height="24" viewBox="0 0 24 24" width="24">
      <path d="M0 0h24v24H0z" fill="none"/>
      <path d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"/>
    </svg>
    `
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NotifyServiceConfig {
    constructor() {
        /**
         * 容器，默认：`document.body`
         */
        this.container = null;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NotifyService {
    /**
     * @param {?} resolver
     * @param {?} applicationRef
     * @param {?} injector
     * @param {?} config
     */
    constructor(resolver, applicationRef, injector, config) {
        this.resolver = resolver;
        this.applicationRef = applicationRef;
        this.injector = injector;
        this.config = config;
        this.emitter = new Subject();
        this.icons = defaultIcons;
        if (config) {
            this.container = config.container;
        }
    }
    /**
     * @param {?} notify
     * @param {?=} to
     * @return {?}
     */
    set(notify, to = true) {
        if (!this.notifyHolderComponent) {
            this.notifyHolderComponent = this.createNotifyHolder();
            this.updateSetting({});
        }
        if (to) {
            notify = Object.assign({}, this.config.notify, notify);
            notify.id = notify.id
                ? notify.id
                : Math.random()
                    .toString(36)
                    .substring(3);
            notify.click = new EventEmitter();
        }
        this.emitter.next({ command: 'set', notify: notify, add: to });
        return notify;
    }
    /**
     * @param {?} options
     * @return {?}
     */
    updateSetting(options) {
        this.emitter.next({
            command: 'options',
            holderOptions: Object.assign({}, this.config.options, options),
        });
    }
    /**
     * @return {?}
     */
    createNotifyHolder() {
        const /** @type {?} */ factory = this.resolver.resolveComponentFactory(NotifyHolderComponent), /** @type {?} */
        ref = factory.create(this.injector), /** @type {?} */
        rootNode = /** @type {?} */ ((/** @type {?} */ (ref.hostView))
            .rootNodes[0]);
        if (!this.container) {
            this.container = document.body;
        }
        this.applicationRef.attachView(ref.hostView);
        ref.onDestroy(() => {
            this.applicationRef.detachView(ref.hostView);
        });
        this.container.appendChild(rootNode);
        return ref.instance;
    }
    /**
     * @return {?}
     */
    getChangeEmitter() {
        return this.emitter;
    }
    /**
     * @param {?} title
     * @param {?=} content
     * @param {?=} override
     * @return {?}
     */
    success(title, content, override) {
        return this.set(Object.assign({}, override, {
            title: title,
            content: content || '',
            type: 'success',
            icon: this.icons.success,
        }));
    }
    /**
     * @param {?} title
     * @param {?=} content
     * @param {?=} override
     * @return {?}
     */
    error(title, content, override) {
        return this.set(Object.assign({}, override, {
            title: title,
            content: content || '',
            type: 'error',
            icon: this.icons.error,
        }));
    }
    /**
     * @param {?} title
     * @param {?=} content
     * @param {?=} override
     * @return {?}
     */
    alert(title, content, override) {
        return this.set(Object.assign({}, override, {
            title: title,
            content: content || '',
            type: 'alert',
            icon: this.icons.alert,
        }));
    }
    /**
     * @param {?} title
     * @param {?=} content
     * @param {?=} override
     * @return {?}
     */
    info(title, content, override) {
        return this.set(Object.assign({}, override, {
            title: title,
            content: content || '',
            type: 'info',
            icon: this.icons.info,
        }));
    }
    /**
     * @param {?} title
     * @param {?} content
     * @param {?} type
     * @param {?=} override
     * @return {?}
     */
    bare(title, content, type, override) {
        return this.set(Object.assign({}, override, {
            title: title,
            content: content || '',
            type: type,
        }));
    }
    /**
     * @param {?} title
     * @param {?} content
     * @param {?} html
     * @param {?=} type
     * @param {?=} override
     * @return {?}
     */
    html(title, content, html, type, override) {
        return this.set(Object.assign({}, override, {
            title: title,
            content: content || '',
            html: html,
            type: type,
        }));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    remove(id) {
        this.emitter.next({ command: 'remove', id: id });
    }
    /**
     * @return {?}
     */
    clear() {
        this.emitter.next({ command: 'clear' });
    }
}
NotifyService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NotifyService.ctorParameters = () => [
    { type: ComponentFactoryResolver, },
    { type: ApplicationRef, },
    { type: Injector, },
    { type: NotifyServiceConfig, decorators: [{ type: Optional },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NotifyHolderComponent {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.notifies = [];
        this.styles = {};
        this.listener = injector
            .get(NotifyService)
            .getChangeEmitter()
            .subscribe(item => {
            switch (item.command) {
                case 'clear':
                    this.notifies = [];
                    break;
                case 'remove':
                    this.remove(/** @type {?} */ ((item.id)));
                    break;
                case 'set':
                    if (item.add)
                        this.add(/** @type {?} */ ((item.notify)));
                    else
                        this.defaultBehavior(item);
                    break;
                case 'options':
                    this.options = Object.assign({
                        lastOnBottom: true,
                        minWidth: 300,
                        maxWidth: 300,
                        maxStack: 8,
                        position: ['right', 'bottom'],
                        offset: [20, 20],
                        zIndex: 1031,
                    }, /** @type {?} */ ((item.holderOptions)));
                    this.styles = {
                        'z-index': this.options.zIndex,
                        'min-width.px': this.options.minWidth,
                        'max-width.px': this.options.maxWidth,
                    };
                    this.styles[this.options.position[0] + '.px'] = this.options.offset[0];
                    this.styles[this.options.position[1] + '.px'] = this.options.offset[1];
                    break;
                default:
                    this.defaultBehavior(item);
                    break;
            }
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @param {?} value
     * @return {?}
     */
    defaultBehavior(value) {
        this.notifies.splice(this.notifies.indexOf(value.notify), 1);
        if (value.notify.onDestroy || this.options.onDestroy) {
            const /** @type {?} */ emitItem = this.getEmit(value.notify, false);
            if (value.notify.onDestroy)
                value.notify.onDestroy(emitItem);
            if (this.options.onDestroy)
                this.options.onDestroy(emitItem);
        }
    }
    /**
     * @param {?} notify
     * @param {?} to
     * @return {?}
     */
    getEmit(notify, to) {
        const /** @type {?} */ res = {
            createdOn: notify.createdOn,
            id: notify.id,
            type: notify.type,
            icon: notify.icon,
            title: notify.title,
            content: notify.content,
            html: notify.html,
        };
        if (!to)
            res.destroyedOn = new Date();
        return res;
    }
    /**
     * @param {?} id
     * @return {?}
     */
    remove(id) {
        let /** @type {?} */ index = 0, /** @type {?} */
        item = null, /** @type {?} */
        allow = false;
        this.notifies.forEach((notify, idx) => {
            if (notify.id === id) {
                index = idx;
                allow = true;
                
            }
        });
        if (allow) {
            this.notifies.splice(index, 1);
        }
    }
    /**
     * @param {?} item
     * @return {?}
     */
    add(item) {
        item = Object.assign({
            type: 'success',
            timeout: 1000 * 3,
            pauseOnHover: true,
            progress: false,
            clickToClose: true,
            theme: 'default',
            animate_in: 'notify-fade-in',
            animate_out: 'notify-fade-out',
            createdOn: new Date(),
        }, item);
        if (!item.html) {
            switch (item.theme) {
                case 'bootstrap':
                    item.html = `<div class="alert alert-{bstype}" role="alert"><strong>{title}</strong> {content}</div>`;
                    break;
                default:
                    item.html = `<div class="notify-title">{title}</div><div class="notify-content">{content}</div><div class="notify-icon">{icon}</div>`;
                    break;
            }
        }
        if (this.options.lastOnBottom) {
            if (this.notifies.length >= this.options.maxStack)
                this.notifies.splice(0, 1);
            this.notifies.push(item);
        }
        else {
            if (this.notifies.length >= this.options.maxStack)
                this.notifies.splice(this.notifies.length - 1, 1);
            this.notifies.splice(0, 0, item);
        }
        if (item.onCreate || this.options.onCreate) {
            const /** @type {?} */ emitItem = this.getEmit(item, true);
            if (item.onCreate)
                item.onCreate(emitItem);
            if (this.options.onCreate)
                this.options.onCreate(emitItem);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.listener)
            this.listener.unsubscribe();
    }
}
NotifyHolderComponent.decorators = [
    { type: Component, args: [{
                selector: 'notify-holder',
                template: `<div class="notifies" [ngStyle]="styles" [ngClass]="options.position" [class]="options.className"> <notify *ngFor="let n of notifies; let i = index" [item]="n"></notify> </div> `,
                styles: [`.notifies {
  position: fixed;
}
.notify {
  width: 100%;
  box-sizing: border-box;
  position: relative;
  margin-bottom: 10px;
  color: #fff;
  cursor: pointer;
  transition: all 0.5s;
}
.notify-theme-default {
  padding: 10px 20px;
}
.notify-theme-default.notify-alert {
  background: #ffdb5b;
}
.notify-theme-default.notify-alert .notify-progress span {
  background: #edc242;
}
.notify-theme-default.notify-success {
  background: #8bc34a;
}
.notify-theme-default.notify-success .notify-progress span {
  background: #689f38;
}
.notify-theme-default.notify-error {
  background: #f44336;
}
.notify-theme-default.notify-error .notify-progress span {
  background: #d32f2f;
}
.notify-theme-default.notify-info {
  background: #03a9f4;
}
.notify-theme-default.notify-info .notify-progress span {
  background: #0288d1;
}
.notify-theme-bootstrap.notify-alert .notify-progress span {
  background: #edc242;
}
.notify-theme-bootstrap.notify-success .notify-progress span {
  background: #689f38;
}
.notify-theme-bootstrap.notify-error .notify-progress span {
  background: #d32f2f;
}
.notify-theme-bootstrap.notify-info .notify-progress span {
  background: #0288d1;
}
.notify-title {
  margin: 0;
  padding: 0;
  line-height: 30px;
  font-size: 16px;
  font-weight: bold;
}
.notify-content {
  margin: 0;
  font-size: 14px;
  padding: 0 50px 0 0;
  line-height: 20px;
}
.notify-icon {
  position: absolute;
  box-sizing: border-box;
  top: 0;
  right: 0;
  height: 100%;
  max-width: 70px;
  width: 100%;
}
.notify-icon svg {
  width: 100%;
  height: 100%;
  padding: 10px;
  fill: #fff;
}
.notify-icon > * {
  max-width: 70px;
}
.notify-progress {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 5px;
}
.notify-progress span {
  float: left;
  height: 100%;
}
.notify-rtl {
  direction: rtl;
}
.notify-rtl .notify-title {
  padding-left: 50px;
}
.notify-rtl .notify-content {
  padding: 0 0 0 50px;
}
.notify-rtl .notify-icon {
  left: 0;
  right: auto;
}
.notify-fade-in {
  animation-duration: 1s;
  animation-fill-mode: both;
  -webkit-animation-name: fadeIn;
  animation-name: fadeIn;
}
.notify-fade-out {
  animation-duration: 1s;
  animation-fill-mode: both;
  -webkit-animation-name: fadeOut;
  animation-name: fadeOut;
}
@-webkit-keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
@keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
@-webkit-keyframes fadeOut {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}
@keyframes fadeOut {
  0% {
    opacity: 1;
  }
  100% {
    opacity: 0;
  }
}
`],
                encapsulation: ViewEncapsulation.None,
                preserveWhitespaces: false
            },] },
];
/** @nocollapse */
NotifyHolderComponent.ctorParameters = () => [
    { type: Injector, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NotifyComponent {
    /**
     * @param {?} sanitizer
     * @param {?} zone
     * @param {?} notifyService
     */
    constructor(sanitizer, zone, notifyService) {
        this.sanitizer = sanitizer;
        this.zone = zone;
        this.notifyService = notifyService;
        this.classes = {};
        this.stopTime = false;
        this.count = 0;
        this.progressWidth = 0;
        this.instance = () => {
            this.zone.runOutsideAngular(() => {
                this.zone.run(() => this.diff = (new Date().getTime() - this.start) - (this.count * this.speed));
                if (this.count++ >= this.steps) {
                    this.zone.run(() => {
                        this.remove();
                    });
                }
                else if (!this.stopTime) {
                    if (this.item.progress)
                        this.zone.run(() => this.progressWidth += 100 / this.steps);
                    this.timer = setTimeout(this.instance, (this.speed - this.diff));
                }
            });
        };
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        switch (this.item.theme) {
            case 'bootstrap':
                switch (this.item.type) {
                    case 'alert':
                        this.item["bstype"] = 'warning';
                        break;
                    case 'error':
                        this.item["bstype"] = 'danger';
                        break;
                    default:
                        this.item["bstype"] = this.item.type;
                        break;
                }
                break;
        }
        this.html = this.sanitizer.bypassSecurityTrustHtml(this.item.html.replace(/\{([a-z]+)\}/g, (full, key) => {
            return this.item[key] || '';
        }));
        this.classes['notify-theme-' + this.item.theme] = true;
        this.classes['notify-' + this.item.type] = true;
        if (this.item.rtl === true)
            this.classes['notify-rtl'] = true;
        if (this.item.animate_in) {
            this.classes['animated'] = true;
            this.classes[this.item.animate_in] = true;
        }
        if (this.item.className) {
            this.classes[this.item.className] = true;
        }
        if (this.item.timeout > 0)
            this.startTimeout();
    }
    /**
     * @return {?}
     */
    startTimeout() {
        this.steps = this.item.timeout / 10;
        this.speed = this.item.timeout / this.steps;
        this.start = new Date().getTime();
        this.zone.runOutsideAngular(() => this.timer = setTimeout(this.instance, this.speed));
    }
    /**
     * @return {?}
     */
    onEnter() {
        if (this.item.pauseOnHover) {
            this.stopTime = true;
        }
    }
    /**
     * @return {?}
     */
    onLeave() {
        if (this.item.pauseOnHover) {
            this.stopTime = false;
            setTimeout(this.instance, (this.speed - this.diff));
        }
    }
    /**
     * @param {?} $e
     * @return {?}
     */
    onClick($e) {
        /** @type {?} */ ((this.item.click)).emit($e);
        if (this.item.clickToClose) {
            this.remove();
        }
    }
    /**
     * @return {?}
     */
    remove() {
        if (this.item.animate_out) {
            this.classes[this.item.animate_in] = false;
            this.classes[this.item.animate_out] = true;
            this.zone.runOutsideAngular(() => {
                setTimeout(() => {
                    this.zone.run(() => this.notifyService.set(this.item, false));
                }, 500);
            });
        }
        else {
            this.notifyService.set(this.item, false);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        clearTimeout(this.timer);
    }
}
NotifyComponent.decorators = [
    { type: Component, args: [{
                selector: 'notify',
                template: `<div class="notify" (click)="onClick($event)" [ngClass]="classes" [class]="animate" (mouseenter)="onEnter()" (mouseleave)="onLeave()"> <div [innerHTML]="html"></div> <div class="notify-progress" *ngIf="item.progress"> <span [ngStyle]="{'width': progressWidth + '%'}"></span> </div> </div> `,
                preserveWhitespaces: false
            },] },
];
/** @nocollapse */
NotifyComponent.ctorParameters = () => [
    { type: DomSanitizer, },
    { type: NgZone, },
    { type: NotifyService, },
];
NotifyComponent.propDecorators = {
    "item": [{ type: Input },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NotifyPushService {
    constructor() {
        this.permission = this.isSupported() ? Notification.permission : 'denied';
    }
    /**
     * @return {?}
     */
    requestPermission() {
        if (this.isSupported())
            Notification.requestPermission((status) => (this.permission = status));
    }
    /**
     * @return {?}
     */
    isSupported() {
        return 'Notification' in window;
    }
    /**
     * @param {?} title
     * @param {?=} options
     * @return {?}
     */
    create(title, options) {
        return new Observable((obs) => {
            if (!this.isSupported()) {
                obs.error('不支持');
                obs.complete();
            }
            if (this.permission !== 'granted') {
                obs.error(`用户未授权`);
                obs.complete();
            }
            const /** @type {?} */ n = new Notification(title, options);
            n.onshow = (e) => obs.next({ notification: n, event: e });
            n.onclick = (e) => obs.next({ notification: n, event: e });
            n.onerror = (e) => obs.error({ notification: n, event: e });
            n.onclose = () => obs.complete();
        });
    }
}
NotifyPushService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
NotifyPushService.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NotifyModule {
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            ngModule: NotifyModule,
            providers: [{ provide: NotifyServiceConfig, useValue: config }],
        };
    }
}
NotifyModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                declarations: [NotifyHolderComponent, NotifyComponent],
                providers: [NotifyService, NotifyPushService],
                entryComponents: [NotifyHolderComponent, NotifyComponent],
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { defaultIcons, NotifyHolderComponent, NotifyComponent, NotifyServiceConfig, NotifyService, NotifyPushService, NotifyModule };
