// @vendure/admin-ui/esm2022/core/shared/components/modal-dialog/dialog-buttons.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./modal-dialog.component";
/**
 * A helper directive used to correctly embed the modal buttons in the {@link ModalDialogComponent}.
 */
export class DialogButtonsDirective {
    constructor(modal, templateRef) {
        this.modal = modal;
        this.templateRef = templateRef;
    }
    ngOnInit() {
        // setTimeout due to https://github.com/angular/angular/issues/15634
        setTimeout(() => this.modal.registerButtonsTemplate(this.templateRef));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DialogButtonsDirective, deps: [{ token: i1.ModalDialogComponent }, { token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.2", type: DialogButtonsDirective, selector: "[vdrDialogButtons]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DialogButtonsDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[vdrDialogButtons]' }]
        }], ctorParameters: function () { return [{ type: i1.ModalDialogComponent }, { type: i0.TemplateRef }]; } });
