// @vendure/admin-ui/esm2022/core/shared/components/modal-dialog/dialog-component-outlet.component.mjs
import { Component, EventEmitter, Input, Output } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * A helper component used to embed a component instance into the {@link ModalDialogComponent}
 */
export class DialogComponentOutletComponent {
    constructor(viewContainerRef) {
        this.viewContainerRef = viewContainerRef;
        this.create = new EventEmitter();
    }
    ngOnInit() {
        const componentRef = this.viewContainerRef.createComponent(this.component);
        this.create.emit(componentRef.instance);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DialogComponentOutletComponent, deps: [{ token: i0.ViewContainerRef }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.2", type: DialogComponentOutletComponent, selector: "vdr-dialog-component-outlet", inputs: { component: "component" }, outputs: { create: "create" }, ngImport: i0, template: ``, isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DialogComponentOutletComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'vdr-dialog-component-outlet',
                    template: ``,
                }]
        }], ctorParameters: function () { return [{ type: i0.ViewContainerRef }]; }, propDecorators: { component: [{
                type: Input
            }], create: [{
                type: Output
            }] } });
