// @vendure/admin-ui/esm2022/react/register-react-custom-detail-component.mjs
import { APP_INITIALIZER } from '@angular/core';
import { CustomDetailComponentService } from '@vendure/admin-ui/core';
import { REACT_CUSTOM_DETAIL_COMPONENT_OPTIONS, ReactCustomDetailComponent, } from './components/react-custom-detail.component';
/**
 * @description
 * Registers a React component to be rendered in a detail page in the given location.
 * Components used as custom detail components can make use of the {@link useDetailComponentData} hook.
 *
 * @docsCategory react-extensions
 */
export function registerReactCustomDetailComponent(config) {
    return {
        provide: APP_INITIALIZER,
        multi: true,
        useFactory: (customDetailComponentService) => () => {
            customDetailComponentService.registerCustomDetailComponent({
                component: ReactCustomDetailComponent,
                locationId: config.locationId,
                providers: [
                    {
                        provide: REACT_CUSTOM_DETAIL_COMPONENT_OPTIONS,
                        useValue: {
                            component: config.component,
                            props: config.props,
                        },
                    },
                ],
            });
        },
        deps: [CustomDetailComponentService],
    };
}
