// @vendure/admin-ui/esm2022/order/components/order-data-table/order-total-column.component.mjs
import { Component } from '@angular/core';
import { DataTable2ColumnComponent } from '@vendure/admin-ui/core';
import * as i0 from "@angular/core";
export class OrderTotalColumnComponent extends DataTable2ColumnComponent {
    constructor() {
        super(...arguments);
        this.orderable = false;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: OrderTotalColumnComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.2", type: OrderTotalColumnComponent, selector: "vdr-order-total-column", exportAs: ["row"], usesInheritance: true, ngImport: i0, template: ``, isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: OrderTotalColumnComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'vdr-order-total-column',
                    template: ``,
                    exportAs: 'row',
                }]
        }] });
