// @vendure/admin-ui/esm2022/core/shared/components/data-table/data-table-column.component.mjs
import { Component, Input, TemplateRef, ViewChild } from '@angular/core';
import * as i0 from "@angular/core";
export class DataTableColumnComponent {
    constructor() {
        /**
         * When set to true, this column will expand to use avaiable width
         */
        this.expand = false;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DataTableColumnComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.2", type: DataTableColumnComponent, selector: "vdr-dt-column", inputs: { expand: "expand" }, viewQueries: [{ propertyName: "template", first: true, predicate: TemplateRef, descendants: true, static: true }], ngImport: i0, template: `
        <ng-template><ng-content></ng-content></ng-template>
    `, isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.2", ngImport: i0, type: DataTableColumnComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'vdr-dt-column',
                    template: `
        <ng-template><ng-content></ng-content></ng-template>
    `,
                }]
        }], propDecorators: { expand: [{
                type: Input
            }], template: [{
                type: ViewChild,
                args: [TemplateRef, { static: true }]
            }] } });
